bpf/tests/ipsec_from_host_test.c

```diff
@@ -0,0 +1,321 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#include "common.h"
+#include <bpf/ctx/skb.h>
+#include "pktgen.h"
+#define ROUTER_IP
+#include "config_replacement.h"
+#undef ROUTER_IP
+
+#define NODE_ID 2333
+#define ENCRYPT_KEY 3
+#define ENABLE_IPV4
+#define ENABLE_IPV6
+#define ENABLE_IPSEC
+#define TUNNEL_MODE
+#define HAVE_ENCAP
+#define ENCAP_IFINDEX 4
+#define SECCTX_FROM_IPCACHE 1
+
+#define skb_set_tunnel_key mock_skb_set_tunnel_key
+#define ctx_redirect mock_ctx_redirect
+
+int mock_skb_set_tunnel_key(__maybe_unused struct __sk_buff *skb,
+			    const struct bpf_tunnel_key *from,
+			    __maybe_unused __u32 size,
+			    __maybe_unused __u32 flags)
+{
+	/* 0xfffff is the default SECLABEL */
+	if (from->tunnel_id != 0xfffff)
+		return -1;
+	if (from->local_ipv4 != 0)
+		return -2;
+	if (from->remote_ipv4 != bpf_htonl(v4_node_two))
+		return -3;
+	return 0;
+}
+
+int mock_ctx_redirect(const struct __sk_buff *ctx __maybe_unused, int ifindex, __u32 flags)
+{
+	if (ifindex != ENCAP_IFINDEX)
+		return -1;
+	if (flags != 0)
+		return -2;
+	return CTX_ACT_REDIRECT;
+}
+
+#include "bpf_host.c"
+
+#define FROM_HOST 0
+#define ESP_SEQUENCE 69865
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 1);
+	__array(values, int());
+} entry_call_map __section(".maps") = {
+	.values = {
+		[FROM_HOST] = &cil_from_host,
+	},
+};
+
+PKTGEN("tc", "ipv4_ipsec_from_host")
+int ipv4_ipsec_from_host_pktgen(struct __ctx_buff *ctx)
+{
+	struct pktgen builder;
+	struct ethhdr *l2;
+	struct iphdr *l3;
+	struct ip_esp_hdr *l4;
+	void *data;
+
+	pktgen__init(&builder, ctx);
+
+	l2 = pktgen__push_ethhdr(&builder);
+	if (!l2)
+		return TEST_ERROR;
+	ethhdr__set_macs(l2, (__u8 *)mac_one, (__u8 *)mac_two);
+
+	l3 = pktgen__push_default_iphdr(&builder);
+	if (!l3)
+		return TEST_ERROR;
+	l3->saddr = v4_pod_one;
+	l3->daddr = v4_pod_two;
+
+	l4 = pktgen__push_esphdr(&builder);
+	if (!l4)
+		return TEST_ERROR;
+	if ((void *)l4 + sizeof(struct ip_esp_hdr) > ctx_data_end(builder.ctx))
+		return 0;
```

Comment:
I would have expected that `pktgen__push_esphdr()` handles this sort of check :thinking: